java-checks/src/main/java/org/sonar/java/checks/helpers/IntersectAutomataChecker.java

```diff
@@ -31,13 +35,20 @@ protected boolean neutralAnswer() {
 
   @Override
   protected boolean checkAuto1AndAuto2Successors(SubAutomaton auto1, SubAutomaton auto2, boolean defaultAnswer, boolean hasConsumedInput) {
-    SimplifiedRegexCharacterClass characterClass1 = SimplifiedRegexCharacterClass.of(auto1.start);
-    SimplifiedRegexCharacterClass characterClass2 = SimplifiedRegexCharacterClass.of(auto2.start);
-    return ((characterClass1 != null) && (characterClass2 != null)) ?
-      (characterClass1.intersects(characterClass2, defaultAnswer) &&
-        auto1.anySuccessorMatch(successor1 -> auto2.anySuccessorMatch(successor2 ->
-          check(successor1, successor2, true)))) :
-      defaultAnswer;
+    boolean start1IntersectsStart2;
+    if (auto1.incomingTransitionType() == BOUNDARY && auto2.incomingTransitionType() == BOUNDARY) {
+      start1IntersectsStart2 = ((BoundaryTree) auto1.start).type() == ((BoundaryTree) auto2.start).type();
```

Comment:
We should return `defaultAnswer`, not `false`, when the boundaries have different types because they could still intersect. Alternatively we could check the intersection of different boundary types like this:

The start of line/input boundaries intersect each other as well as the word and grapheme boundaries.
The end of line/input (with or without final terminator) boundaries intersect each other as well as the word and grapheme boundaries.
The non-word boundary intersects only itself.